### ISSUE TITLE
TreeGrid: expand error when hierarchy level more than 2 and sub items were already expanded

### ISSUE THREAD

**Issue by sirbris** [vaadin/flow-components#3355]
### Description

When having a TreeGrid with a deep hierarchy the new eager fetch mode does seem to end up in a loop when trying to open a parent with a child which is also a parent which was previously opened.

### Expected outcome

Expanding a parent should open without endless loading

### Minimal reproducible example

TreeGrid with a TreeDataProvider like this:
A
 - A1
   - B1
     - C1
     - C2
     - C3
   - B2
   - B3
 - A2

B
C

In my case there are some component renderers used.

### Steps to reproduce

1. Open Item A, A1 and B1
2. Close A
3. Open A

### Environment

Vaadin version(s): 23.1.0
OS: Chrome OS

### Browsers

Chrome

**Comment by sirbris** [comment]
After I expand/collapse some items wich themselves were expanded before, I see in this simple example without any renderers following error in the debug console... hope that helps, if you have any questions, please don't hesitate.
<img width="1177" alt="tree_grid_expand_error" src="https://user-images.githubusercontent.com/664179/174243745-67734a88-ca23-4418-ba84-76a20c3688a7.png">

**Comment by tomivirkki** [comment]
Hi @sirbris 

I tried to reproduce this according to your instructions but it works fine for me. Tried with server push on and off. Please provide an example view I could use to debug the problem you're facing.

https://user-images.githubusercontent.com/1222264/174254550-6f7fe423-6415-4e65-9784-2f9e393ddbf5.mp4

**Comment by sirbris** [comment]
Thanks for looking into this... I'll provide a more detailed example shortly... Could this be Jetty 10 related?

**Comment by tomivirkki** [comment]
> Could this be Jetty 10 related?

Not sure what it could be related to. I tried with Jetty 10.0.9 but still can't reproduce the issue

**Comment by sirbris** [comment]
In the video I kind of stress tested the TreeGrid with this simplified example, but in a more complex scenario it happens way more frequently:
`
       TreeGrid<String> tree = new TreeGrid<>(String.class);
        tree.removeAllColumns();
        tree.addHierarchyColumn(item -> item).setHeader("Column");
        tree.addColumn(item -> item.length() + "");
        tree.addComponentColumn(item -> VaadinIcon.ACCESSIBILITY.create());
        TreeData<String> treeData = new TreeData<>();
        TreeDataProvider<String> provider = new TreeDataProvider<>(treeData);
        tree.setDataProvider(provider);
        treeData.addRootItems("A", "B", "C");
        treeData.addItems("A", "A1", "A2");
        treeData.addItems("A1", "C1", "C2");
        treeData.addItems("C1", "D1", "D2");
        tree.getDataProvider().refreshAll();
        tree.getDataCommunicator().reset();

        add(tree);
`
[treeGridExpandError.mov.zip](https://github.com/vaadin/flow-components/files/8926900/treeGridExpandError.mov.zip)

**Comment by sirbris** [comment]
JFYI: I could replicate the same behavior with Vaadin 14.8.13 as well

**Comment by Kaka0Express** [comment]
I can add some further information:
Tree for testing was a Item with two children:


![tree](https://user-images.githubusercontent.com/107686442/174292586-39a02d60-5369-468c-9ebf-f5e1e0de0e7b.PNG)


I activated the debuglevel in the jetty-websocket:
org.eclipse.jetty.websocket.LEVEL=DEBUG

If you expand "A" you get following three Websocket Log entries:
2022-06-17 08:15:38,240 DEBUG [qtp192620230-328] WebSocketMappings: WebSocket Negotiated detected on / for endpoint CreatorNegotiator@31a20dc0{JsrCreator[org.atmosphere.container.JSR356Endpoint],org.eclipse.jetty.websocket.javax.server.internal.JavaxWebSocketServerFrameHandlerFactory@695bd33}
2022-06-17 08:15:38,241 DEBUG [qtp192620230-328] AbstractHandshaker: not upgraded method!=GET Request(POST http://xxx.xxx.xxx.xxx/_flow/?v-r=uidl&v-uiId=0)@18689f2e
2022-06-17 08:15:38,241 DEBUG [qtp192620230-328] AbstractHandshaker: not upgraded method!=GET Request(POST http://xxx.xxx.xxx.xxx/_flow/?v-r=uidl&v-uiId=0)@18689f2e
2022-06-17 08:15:38,362 DEBUG [qtp192620230-330] WebSocketMappings: WebSocket Negotiated detected on / for endpoint CreatorNegotiator@31a20dc0{JsrCreator[org.atmosphere.container.JSR356Endpoint],org.eclipse.jetty.websocket.javax.server.internal.JavaxWebSocketServerFrameHandlerFactory@695bd33}
2022-06-17 08:15:38,362 DEBUG [qtp192620230-330] AbstractHandshaker: not upgraded method!=GET Request(POST http://xxx.xxx.xxx.xxx/_flow/?v-r=uidl&v-uiId=0)@18689f2e
2022-06-17 08:15:38,362 DEBUG [qtp192620230-330] AbstractHandshaker: not upgraded method!=GET Request(POST http://xxx.xxx.xxx.xxx/_flow/?v-r=uidl&v-uiId=0)@18689f2e
2022-06-17 08:15:38,435 DEBUG [qtp192620230-328] WebSocketMappings: WebSocket Negotiated detected on / for endpoint CreatorNegotiator@31a20dc0{JsrCreator[org.atmosphere.container.JSR356Endpoint],org.eclipse.jetty.websocket.javax.server.internal.JavaxWebSocketServerFrameHandlerFactory@695bd33}
2022-06-17 08:15:38,436 DEBUG [qtp192620230-328] AbstractHandshaker: not upgraded method!=GET Request(POST http://xxx.xxx.xxx.xxx/_flow/?v-r=uidl&v-uiId=0)@18689f2e
2022-06-17 08:15:38,436 DEBUG [qtp192620230-328] AbstractHandshaker: not upgraded method!=GET Request(POST http://xxx.xxx.xxx.xxx/_flow/?v-r=uidl&v-uiId=0)@18689f2e



If you expand "A" and the subelements are not shown you only get one WebSocket Log entry:
2022-06-17 08:17:18,642 DEBUG [qtp192620230-328] WebSocketMappings: WebSocket Negotiated detected on / for endpoint CreatorNegotiator@31a20dc0{JsrCreator[org.atmosphere.container.JSR356Endpoint],org.eclipse.jetty.websocket.javax.server.internal.JavaxWebSocketServerFrameHandlerFactory@695bd33}
2022-06-17 08:17:18,642 DEBUG [qtp192620230-328] AbstractHandshaker: not upgraded method!=GET Request(POST http://xxx.xxx.xxx.xxx/_flow/?v-r=uidl&v-uiId=0)@18689f2e
2022-06-17 08:17:18,642 DEBUG [qtp192620230-328] AbstractHandshaker: not upgraded method!=GET Request(POST http://xxx.xxx.xxx.xxx/_flow/?v-r=uidl&v-uiId=0)@18689f2e


![failedexpand](https://user-images.githubusercontent.com/107686442/174297655-f457657a-e119-48ea-8369-c3967cd76e17.PNG)

**Comment by tomivirkki** [comment]
> JFYI: I could replicate the same behavior with Vaadin 14.8.13 as well

This kind of suggests that the problem isn't related to eager fetch mode (but TreeGrid in general) since the feature hasn't been backported to V14

**Comment by sirbris** [comment]
@tomivirkki can I provide more information to help figure this one out? I already requested the support to prioritize this issue via Vaadin warranty.

**Comment by sirbris** [comment]
I could even reproduce it with one column, no component renderers whatsoever... it seems to be easier to reproduce in a more complex environment though. According to the Jetty logs from @Kaka0Express I'd guess, that in certain scenarios the expand request gets lost, although the client side is still able to change the toggle icon.

**Comment by sirbris** [comment]
Here another example, where I start to click slowly and then I proceed with almost double click speed. The collapse/expand icon then suggests, that the node state is collapsed but something is off, because the next click toggles the icon but no children are shown.
https://user-images.githubusercontent.com/664179/174543831-86b31703-5094-48a5-a2aa-b1a058af6acf.mov

**Comment by tomivirkki** [comment]
This is probably the same as https://github.com/vaadin/web-components/issues/2391 (the issue should also be in the flow-components repo)

**Comment by sirbris** [comment]
Is there already a possible hack to prevent the double click or some other workaround to avoid this behavior? Thanks in advance!

**Comment by sirbris** [comment]
A small update would be greatly appreciated... we really rely on this issue being fixed, no matter if via new minor release, via patch or via hacky workaround. Thanks in advance for any information about this!

**Comment by sissbruecker** [comment]
@sirbris Can you please elaborate what the impact of this issue is on your application? All the reproductions I have seen in this issue involve double- or rapidly clicking on the toggle button. Does the issue also occur in your app when leaving some amount of time (for example a second) between clicks? Do you expect users to often close a tree node and open it immediately after, and is there a specific use-case around that in your app? This is not about validating your right to prioritize this issue of course, we just want to get more insight into the impact of the problem and possible solutions. Thanks!

As for workarounds: If preventing double-clicks is useful to you, then I can provide the following workaround. Note that technically we are not dealing with a double-click, which is a separate event in JS, rather we are dealing with two clicks following in quick succession. The following JS code snippet will note when a specific tree toggle has been clicked, and prevent subsequent clicks for a specific amount of time (see `preventClickTimeout` variable):
```js
window.Vaadin = window.Vaadin || {};
window.Vaadin.Flow = window.Vaadin.Flow || {};
window.Vaadin.Flow.registerTreeToggleDoubleClickWorkaround = function(grid) {
  const preventClickTimeout = 500; // = 500ms
  const lastToggleClicks = new Map();

  // Register click listener in capture phase
  grid.addEventListener('click', handleClick, true);

  function handleClick(event) {
    // Only apply workaround to tree toggles
    const isTreeToggle = event.target.tagName.toLowerCase() === 'vaadin-grid-tree-toggle';
    if (!isTreeToggle) return;
    // Get last time the toggle was clicked
    const lastClickTime = lastToggleClicks.get(event.target);
    // If there was a previous click, and it's within the prevent click timeout, stop propagation
    const now = new Date().getTime();

    if (lastClickTime && now - lastClickTime < preventClickTimeout) {
      event.stopPropagation();
    } else {
      // Otherwise store last click time in map
      lastToggleClicks.set(event.target, now);
    }
  }
}
```

To use this, add it to a file like `frontend/tree-grid-toggle-workaround.js` and import that by adding a `@JsModule("./tree-grid-toggle-workaround.js")` annotation to the view where you use the tree grid. Then enable the workaround for a grid using:
```java
grid.getElement().executeJs("window.Vaadin.Flow.registerTreeToggleDoubleClickWorkaround($0)", grid.getElement());
```

**Comment by F0rce** [comment]
Some additional information:

the provided workaround seems to work with v14 and some clicks are lost (because of the debounce, which is good). But as soon as I try to include the same provided workaround with v23 nothing happens and the same logic remains.

**Comment by lucaschraml** [comment]
@sirbris @F0rce +1 can reproduce

**Comment by F0rce** [comment]
![grafik](https://user-images.githubusercontent.com/60381251/176918481-2a55a982-c23e-4892-98a4-d0c9180aacf9.png)

after more investigation, I noticed that the clicks (with v23) on the grid are kind of buffered / queued (maybe thats a cause for this whole problem). I changed the provided workaround so that it logs the event which was stopped or if it was stored and as soon as the loading bar appeared at the top of the screen (please see @sirbris attached video https://github.com/vaadin/flow-components/issues/3355#issuecomment-1172381336) and we wait some seconds the browser console was spammed with the click events... UI is still usable, but it seems like it's loading endlessly.

Additionally the console tells me open a new issue as there is a bug... but can't tell why it happens.

**Comment by sirbris** [comment]
Any further findings on this? It's much easier to reproduce this bug with lots of items in the DataProvider and heavy use in general. The weird thing in V23 is the appearing progress or loading bar at the top... this suggests, that the user has to wait for its completion, but it'll stay there until the user clicks around.

**Comment by sissbruecker** [comment]
@sirbris We've identified an issue with multiple expand / collapse requests for the same tree node running in parallel that results in a similar behavior to what is observed in these comments: https://github.com/vaadin/flow-components/issues/3355#issuecomment-1160057115, https://github.com/vaadin/flow-components/issues/3355#issuecomment-1161755196, as well as this issue: https://github.com/vaadin/web-components/issues/2391. We could create a fix for that.

However the issue above is only reproducible by rapid clicking, or significantly throttling the network connection. As you mentioned that for you this also happens for regular usage, which in the context of the previous discussion I understand as not clicking twice on the same tree toggle in a short amount of time, we are still missing a reproduction for that. I spend some time yesterday with some of the test setups in this repository, which contain large nested trees, and could not observe any issues by slowly clicking around.

The only thing I can think of at the moment is that I provide you with a branch / patch for this repository that includes the fix mentioned above, which would allow you to build a snapshot and use that to test it against your application to see if that actually solves your problem.

**Comment by F0rce** [comment]
@sissbruecker thanks for the update,

where could I find the branch that I could build a snapshot. Could you also provide me with small information on how to build that branch ? I was trying some time ago and it seemed super complicatied.

**Comment by sissbruecker** [comment]
The branch is [`fix/treegrid_out_of_sync`](https://github.com/vaadin/flow-components/tree/fix/treegrid_out_of_sync).

In order to use this you would:

Clone this repo and checkout the branch:
```
git clone git@github.com:vaadin/flow-components.git
cd flow-components
git pull
git checkout treegrid_out_of_sync
```
Build a snapshot of the Grid module:
```
cd vaadin-grid-flow-parent/vaadin-grid-flow
mvn install -DskipTests=true
```
That will install the snapshot of the Grid in your local Maven repo, using the version `23.2-SNAPSHOT`.

Then modify your application to use the Vaadin version `23.2-SNAPSHOT` and run it. You need to have the Vaadin pre-releases Maven repository declared in your POM for this to work, which should look like this:
```xml
    <repositories>
        <!-- The order of definitions matters. Explicitly defining central here to make sure it has the highest priority. -->

        <!-- Main Maven repository -->
        <repository>
            <id>central</id>
            <url>https://repo.maven.apache.org/maven2</url>
            <snapshots>
                <enabled>false</enabled>
            </snapshots>
        </repository>
        <repository>
            <id>vaadin-prereleases</id>
            <url>
                https://maven.vaadin.com/vaadin-prereleases/
            </url>
        </repository>
        <!-- Repository used by many Vaadin add-ons -->
        <repository>
            <id>Vaadin Directory</id>
            <url>https://maven.vaadin.com/vaadin-addons</url>
            <snapshots>
                <enabled>false</enabled>
            </snapshots>
        </repository>
    </repositories>
```
That should download snapshots for all the other Vaadin artifacts, except for the grid which you already installed manually.

**Comment by sirbris** [comment]
Just to clarify: in order to use the patch we have to move our whole application to the snapshot version of v23.2 or can we still use 23.1.3 with the fixed snapshot version of the grid?
When can we expect the stable release of v23.2.0? Thanks in advance!

**Comment by sirbris** [comment]
So the last couple of days we heavily tested the provided fix (cherry-picked fix in v23.1 branch)... it definitely has gotten better, but it's still reproducible. If there's anything else we can do to further eliminate the broken state after multiple clicks, we'd be happy to hear your suggestions. Thanks in advance!

**Comment by sirbris** [comment]
It generally seems to be related to the expand method... when we move an already expanded item (with also expanded items inside) in another hierarchy via drag and drop, pretty much the same buggy state as rapid clicking on the collapse/expand icon happens and the loading indicator stays active

**Comment by vursen** [comment]
The [upcoming TreeGrid changes](https://github.com/vaadin/flow-components/pull/7676) (will be available in Vaadin 25) seem to resolve this glitch:

https://github.com/user-attachments/assets/c483d792-8556-423d-839f-eacfa098a79a

**Comment by vursen** [comment]
Resolved by https://github.com/vaadin/platform/issues/7843 in Vaadin 25.
